### ISSUE TITLE
Model configuration from GGUF metadata

### ISSUE THREAD

**Issue by anibjoshi** [strata-ai-labs/strata-inference#11]
### Context
GGUF metadata fully describes a model's architecture. We need a config system that reads these KV pairs and produces a typed Rust struct that drives the transformer forward pass. This replaces strata-core's hardcoded constants.

### Architecture Detection
```
general.architecture → "gemma3" | "llama" | "bert" | "nomic-bert" | ...
```

### ModelConfig Struct
```rust
pub struct ModelConfig {
    pub arch: ModelArch,             // Enum: Gemma, LLaMA, BERT, ...
    pub hidden_size: usize,          // {arch}.embedding_length
    pub num_layers: usize,           // {arch}.block_count
    pub num_heads: usize,            // {arch}.attention.head_count
    pub num_kv_heads: usize,         // {arch}.attention.head_count_kv
    pub head_dim: usize,             // hidden_size / num_heads
    pub ffn_hidden: usize,           // {arch}.feed_forward_length
    pub vocab_size: usize,           // {arch}.vocab_size
    pub max_seq_len: usize,          // {arch}.context_length
    pub norm_type: NormType,         // LayerNorm vs RMSNorm
    pub norm_eps: f32,               // epsilon for normalization
    pub activation: Activation,      // GELU vs SiLU vs SwiGLU
    pub position_type: PositionType, // Learned vs RoPE
    pub rope_freq_base: f32,         // {arch}.rope.freq_base (default 10000)
    pub causal: bool,                // {arch}.attention.causal
    pub pooling: PoolingType,        // NONE / MEAN / CLS / LAST
}

pub enum ModelArch { Gemma, Gemma3, LLaMA, BERT, NomicBERT, ... }
pub enum NormType { LayerNorm, RMSNorm }
pub enum Activation { GELU, SiLU, SwiGLU }
pub enum PositionType { Learned, RoPE }
pub enum PoolingType { None, Mean, Cls, Last }
```

### Key Combinations by Architecture
| Arch | Norm | Activation | Position | Causal | Pooling |
|------|------|------------|----------|--------|---------|
| BERT | LayerNorm | GELU | Learned | No | Mean/CLS |
| Gemma3 | RMSNorm | SwiGLU | RoPE | Depends | Mean |
| LLaMA | RMSNorm | SwiGLU | RoPE | Yes | None |

### Acceptance Criteria
- [ ] `ModelConfig::from_gguf(gguf: &GgufFile) -> Result<ModelConfig>`
- [ ] Detect architecture from `general.architecture`
- [ ] Read all dimension parameters with sensible defaults
- [ ] Infer norm_type, activation, position_type from architecture
- [ ] Override defaults with explicit GGUF keys when present
- [ ] Validate consistency (e.g., hidden_size % num_heads == 0)
- [ ] Pretty-print config for logging
- [ ] Test with EmbeddingGemma-300M GGUF and BERT GGUF metadata

### Reference
- llama.cpp: `src/llama-model.cpp` lines 643-2170 (parameter loading switch)
- llama.cpp: `src/llama-arch.h` lines 135-331 (metadata key enum)

### Estimated LOC: ~400-500

**Comment by anibjoshi** [comment]
## Critical Gaps from llama.cpp Evaluation

After comparing our M1-M3 implementation against llama.cpp, here are critical items that must be addressed in the `ModelConfig` implementation:

### New Fields Required in ModelConfig

```rust
pub struct ModelConfig {
    // ... existing fields ...
    
    // NEW — Critical for EmbeddingGemma-300M
    pub attn_logit_softcap: f32,    // {arch}.attn_logit_softcapping (default 0.0 = disabled)
    pub embedding_scale: f32,        // sqrt(hidden_size) for Gemma, 1.0 for others
    pub rope_dim: usize,             // {arch}.rope.dimension_count (default = head_dim)
    pub key_length: Option<usize>,   // {arch}.attention.key_length (may differ from head_dim)
    pub value_length: Option<usize>, // {arch}.attention.value_length

    // NEW — Nice-to-have for broader model support
    pub sliding_window: Option<usize>,  // {arch}.attention.sliding_window
    pub attn_scale: Option<f32>,        // {arch}.attention.scale (custom instead of 1/sqrt(d_k))
}
```

### Metadata Keys Comparison (us vs llama.cpp)

| Key | Status | Impact |
|-----|--------|--------|
| `{arch}.attn_logit_softcapping` | **MISSING — CRITICAL** | Gemma3 requires `soft_cap * tanh(logits / soft_cap)` on attention scores |
| `{arch}.embedding_scale` | **MISSING — CRITICAL** | Gemma scales embeddings by `sqrt(hidden_dim)` after lookup |
| `{arch}.rope.dimension_count` | **MISSING — HIGH** | Controls partial RoPE (only first N dims rotated). Most models use full head_dim, but must be read |
| `{arch}.attention.key_length` | MISSING | Per-head key dim may differ from `hidden_size / n_heads` |
| `{arch}.attention.value_length` | MISSING | Per-head value dim |
| `{arch}.pooling_type` | In spec but needs implementation | 0=NONE, 1=MEAN, 2=CLS, 3=LAST |
| `{arch}.attention.causal` | In spec but needs implementation | Determines bidirectional vs causal mask |
| `{arch}.attention.sliding_window` | MISSING | Sliding window attention for efficiency |
| `{arch}.rope.scaling_type` | MISSING (not needed for target) | Linear/YaRN/NTK scaling for extended context |

### Gemma3-Specific Config

llama.cpp reads these Gemma3-specific parameters that we need:
1. **`attn_logit_softcapping`** — Applied before softmax: `logits = cap * tanh(logits / cap)`. Without this, attention scores can blow up and produce wrong embeddings.
2. **Embedding scaling** — After token embedding lookup, multiply by `sqrt(hidden_dim)`. llama.cpp does this in the model builder. Without it, the scale of activations is wrong from layer 1.
3. **Pooling type** — Must read from metadata to know whether to use MEAN pooling (typical for embedding models) vs CLS vs LAST.

### Decision: GGUF v2 Handling

Our parser currently accepts GGUF v2 but parses it with v3 semantics (u64 string lengths). V2 uses u32 string lengths, so v2 files would be silently misparsed. **Decision needed:** either reject v2 explicitly or add v2-specific parsing. Recommendation: reject v2 with a clear error since all current models use v3.

### Architecture Detection Table (from llama.cpp)

llama.cpp supports 130+ architectures. For our target, we need:

| `general.architecture` | Norm | Activation | Position | Embedding Scale | Logit Softcap |
|------------------------|------|------------|----------|-----------------|---------------|
| `gemma3` | RMSNorm | SwiGLU | RoPE | `sqrt(hidden_size)` | Yes (read from metadata) |
| `gemma2` | RMSNorm | SwiGLU | RoPE | `sqrt(hidden_size)` | Yes |
| `llama` | RMSNorm | SwiGLU | RoPE | 1.0 | No |
| `bert` | LayerNorm | GELU | Learned | 1.0 | No |
